h2o-core/src/main/java/hex/ScoreKeeper.java

```diff
@@ -111,30 +111,133 @@ else if (m instanceof ModelMetricsMultinomial) {
       _custom_metric =  m._custom_metric.value;
   }
 
-  public enum StoppingMetric { AUTO, deviance, logloss, MSE, RMSE,MAE,RMSLE, AUC, lift_top_group, misclassification, mean_per_class_error, custom, custom_increasing} //, r2}
-  static boolean moreIsBetter(StoppingMetric criterion) {
-    return (criterion == StoppingMetric.AUC || criterion == StoppingMetric.lift_top_group || criterion == StoppingMetric.custom_increasing); // || criterion == StoppingMetric.r2);
+  public interface IStoppingMetric {
+    int direction();
+    boolean isLowerBoundBy0();
+    IConvergenceStrategy getConvergenceStrategy();
+    double metricValue(ScoreKeeper sk);
   }
-  // only for those metrics where "less is better"
-  private static boolean hasLowerBound(StoppingMetric criterion) {
-    return criterion != StoppingMetric.deviance;
+  
+  public enum StoppingMetric implements IStoppingMetric {
+    AUTO(ConvergenceStrategy.AUTO, false), 
+    deviance(ConvergenceStrategy.LESS_IS_BETTER, false),
+    logloss(ConvergenceStrategy.LESS_IS_BETTER, true),
+    MSE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    RMSE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    MAE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    RMSLE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    AUC(ConvergenceStrategy.MORE_IS_BETTER, false),
+    lift_top_group(ConvergenceStrategy.MORE_IS_BETTER, false),
+    misclassification(ConvergenceStrategy.LESS_IS_BETTER, true),
+    mean_per_class_error(ConvergenceStrategy.LESS_IS_BETTER, true),
+    custom(ConvergenceStrategy.LESS_IS_BETTER, false),
+    custom_increasing(ConvergenceStrategy.MORE_IS_BETTER, false),
+    anomaly_score(ConvergenceStrategy.NON_DIRECTIONAL, false);
+
```

Comment:
Why is anomaly_score NON_DIRECTIONAL?  How do you determine if you are doing better or not if you do not know the direction it should go then?

Reply:
Isolation Forest is unsupervised, there is no way to measure if it is doing better or worse. The early stopping is a based on the mean anomaly score stabilizing - that is it stops changing in either direction. It doesn't go "a lot" up or "a lot" down. That is how we tell the algo to stop early.